[improve][fn] Run search connectors in parallel

aymkhalil opened the pull request:
<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://docs.google.com/document/d/1d8Pw6ZbWk-_pCKdOmdvx9rnhPiyuxwq60_TrD68d7BA/edit#heading=h.trs9rsex3xom)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

### Motivation

Improve functions worker bootstrap time with increased number of connectors. Currently all nar files are unpacked during startup to perform validation steps (e.g. if the connector class implements the correct interfaces). For that, the NarClassLoader will unpack all nars to disk sequentially which could delay startup.

On my Machine (Mac M1) I observed an improvement of ~ 3X (from 48 seconds down to 15 seconds). I know the max number of chosen threads (4) is a magic number (I have not seen improvements beyond that number), but if this is the right approach and it is worth it, we can make this number configurable (need guidance on the right config place)

Sample output:
```
2022-09-30T01:22:13,778+0000 [main] INFO  org.apache.pulsar.functions.utils.io.ConnectorUtils - Searching for connectors in /pulsar/./connectors
2022-09-30T01:22:13,779+0000 [main] INFO  org.apache.pulsar.functions.utils.io.ConnectorUtils - Loading 58 connector definitions with a thread pool of size 4
2022-09-30T01:22:13,781+0000 [search-connectors-executor-0] INFO  org.apache.pulsar.common.nar.NarUnpacker - Created directory /tmp/pulsar-nar/pulsar-io-debezium-mssql-2.10.1.8-SNAPSHOT.nar-unpacked
2022-09-30T01:22:13,781+0000 [search-connectors-executor-1] INFO  org.apache.pulsar.common.nar.NarUnpacker - Created directory /tmp/pulsar-nar/pulsar-snowflake-connector-0.1.10.nar-unpacked
2022-09-30T01:22:13,781+0000 [search-connectors-executor-2] INFO  org.apache.pulsar.common.nar.NarUnpacker - Created directory /tmp/pulsar-nar/pulsar-io-mongo-2.10.1.8-SNAPSHOT.nar-unpacked
2022-09-30T01:22:13,784+0000 [search-connectors-executor-3] INFO  org.apache.pulsar.common.nar.NarUnpacker - Created directory /tmp/pulsar-nar/pulsar-io-kinesis-2.10.1.8-SNAPSHOT.nar-unpacked
2022-09-30T01:22:13,895+0000 [search-connectors-executor-2] INFO  org.apache.pulsar.common.nar.NarUnpacker - Extracting /pulsar/./connectors/pulsar-io-mongo-2.10.1.8-SNAPSHOT.nar to /tmp/pulsar-nar/pulsar-io-mongo-2.10.1.8-SNAPSHOT.nar-unpacked/gs7wF7XmmgT1MjvpwSxFfA
```

### Modifications

Just run the connectors search function (which unpacks the nars) in parallel via a CachedThreadPool.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change is already covered by existing tests, such as the unit tests under PulsarFunctionLocalRunTest already covers this path. Also CI for each existing connector will cover the change

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [x] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### Matching PR in forked repository

PR in forked repository: (https://github.com/aymkhalil/pulsar/pull/2)

After opening this PR, the build in apache/pulsar will fail and instructions will
be provided for opening a PR in the PR author's forked repository.

apache/pulsar pull requests should be first tested in your own fork since the 
apache/pulsar CI based on GitHub Actions has constrained resources and quota.
GitHub Actions provides separate quota for pull requests that are executed in 
a forked repository.

The tests will be run in the forked repository until all PR review comments have
been handled, the tests pass and the PR is approved by a reviewer.

-->

Commit:
[improve][fn] Run search connectors in parallel

eolivelli commented on pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/io/ConnectorUtils.java:
```diff
@@ -53,6 +59,8 @@ public class ConnectorUtils {
 
     private static final String PULSAR_IO_SERVICE_NAME = "pulsar-io.yaml";
 
+    private static final Executor cachedThreadPool = Executors.newCachedThreadPool();
```
We have to control the life cycle of this ThreadPool
leaving a static ThreadPool is generally a not a good practice.
In this case we can create the ThreadPool and dispose it in searchForConnectors.

Also I would use a fixed size threadpool, otherwise we have to control on the parallelism

eolivelli replied:
we should also provide a meaningful name to the threads

eolivelli reviewed (changes requested):
the approach looks good to me, I left some comments

Commit:
Use Runtime.getRuntime().availableProcessors() to init the search connectors thread pool

aymkhalil committed:
Update pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/io/ConnectorUtils.java

Co-authored-by: Lari Hotari <lhotari@users.noreply.github.com>

Commit:
Use FutureUtil.waitForAll()

Technoboy- committed:
[improve][common] Add thread name for RateLimiter. (#18013)

lordcheng10 committed:
Avoid unnecessary creation of BitSetRecyclable objects (#17998)

Co-authored-by: leolinchen <leolinchen@tencent.com>

shibd committed:
[refactor][client] Remove useless locks in batch receiving. (#18005)

poorbarcode committed:
[improve][broker] Make CompactedTopicImpl.findStartPointLoop work more efficiently (#17976)

coderzc committed:
[fix][broker] Fix the order of resource close in the InMemoryDelayedDeliveryTracker (#18000)

coderzc committed:
[improve][test] Improve AdminApiTest to reduce the execution time (#17980)

Technoboy- committed:
[fix][broker] Fix system service namespace create internal event topic. (#17867)

lhotari committed:
[improve][docs] Improve docs for persistence policies parameters E, Qw and Qa. (#18003)

* [improve][docs] Improve docs for Bookkeeper E, Qw and Qa.

Technoboy- committed:
[fix][broker] Fix `getPositionAfterN` infinite loop. (#17971)

dave2wave committed:
[improve][misc] Highlight change to threading (#18025)

* [improve][misc] Highlight change to threading 

Whenever a PR changes the threading model we need to make clear that we should track these changes

* Fix typo

poorbarcode committed:
[fix][test] flaky AdminApi2Test.cleanup (#17861)

poorbarcode committed:
[improve][test] Improve KeySharedSubscriptionTest to reduce the execution time (#18027)

Pomelongan committed:
[fix][broker] Update the log print content of createSubscriptions (#18024)

Co-authored-by: huangzegui <huangzegui@didiglobal.com>

poorbarcode committed:
[improve][test] Improve SimpleSchemaTest to reduce the execution time (#18032)

Huanli-Meng committed:
[fix][doc] Fix the grammar error in docs (#18033)

coderzc committed:
[fix][io] Remove unused import statement in the MariadbJdbcAutoSchemaSink (#18031)

cbornet committed:
[fix][connectors] Fix builtin sink transformation on k8s (#18019)

poorbarcode committed:
[fix][test] Fix flaky test AdminApi2Test.resetClusters (#18043)

Commit:
[fix] [pulsar-client] Fix pendingLookupRequestSemaphore leak when channel inactive (#17856)

### Motivation
https://github.com/apache/pulsar/blob/b89c1451551a6bbe681465726906a2e61c9d8a69/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java#L282-L297 
The `pendingLookupRequestSemaphore` will leak when channel inactive. There are `LookUpRequestSemaphore` not released when removing it from `pendingRequests`

### Modifications
We can't easily release the semaphore in `channelInactive`, because there are not only `LookUpRequest`. So release the semaphore when connectionException

### Verifying this change
Add unit test case to cover this change

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [x] `doc-not-needed` 
bug fixs, no need doc

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

Anonymitaet committed:
[improve][doc] Update workflow of submitting release note (#18047)

michaeljmarshall committed:
[cleanup][proxy] Use correct address for HAProxyMessage destination (#16045)

* [fix][proxy] Use correct channel for HAProxyMessage dest address

* Fix implementation after researching HAProxy

### Motivation

When reading through the Proxy code, I noticed that the `remoteAddress` was used where it probably makes sense to use the `localAddress`.

After researching it a bit with the AWS docs https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/enable-proxy-protocol.html and the protocol spec http://www.haproxy.org/download/1.8/doc/proxy-protocol.txt, I think this implementation is correct.

We could look at using the value in the broker logs if we wanted to make it clear which proxy the traffic is ingressing through. For now, I propose we just fix the code in the proxy.

### Modifications

* Replace `inboundChannel.localAddress` with `inboundChannel.localAddress`.
* Fix type checking

- [x] `doc-not-needed`

cbornet committed:
[fix][io] Fix builtin connector/function download filename (#18044)

When using KubernetesRuntime, if uploadBuiltinSinksSources is false (non-default), the originalFileName
is not set for builtin components. But with k8s, we need a filename since the component will be
downloaded. This results in the component currently being downloaded with the download directory as filename.
Also when using a transform function, we use the originalFileName to know if there's a builtin function to download.

coderzc committed:
Bump commons-text to 1.10.0 (#18053)

coderzc committed:
Add link to CPP/Python client to README.md (#18054)

lhotari committed:
[fix][ci] Fix deprecation warnings about set-output (#18048)

Error message is:
The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

eolivelli committed:
Make BookieId work with PulsarRegistrationDriver (second take) (#17922)

* Make BookieId work with PulsarRegistrationDriver (#17762)

* Make BookieId work with PulsarRegistrationDriver

* Switch to MetadataCache

* checkstyle

* Do not execute lookup on MetadataCache in the getBookieServiceInfo caller thread

AlvaroStream committed:
[Improve][doc] Insist about default behaviour for retention. (#17958)

liangyepianzhou committed:
[fix][test] Fix TransactionTest failure due to clean up after class (#17746)

congbobo184 committed:
[fix][test] AdvertisedListenersTest.setup (#17869)

dragonls committed:
[improve][test] Add test case for system topic schema not compatible bug. (#17992)

Co-authored-by: druidliu <druidliu@tencent.com>

coderzc committed:
[improve][test] Improve SubscriptionMessageDispatchThrottlingTest to reduce the execution time (#18039)
